### DISCUSSION TITLE
web-api-google-authentication-example

### DISCUSSION THREAD

**Discussion by giscus** [bloggrammer/DevBlog#12] [Announcements]
# web-api-google-authentication-example

Discover how to seamlessly integrate Google authentication into your ASP.NET Web API. Follow step-by-step instructions, including creating Google client secrets and IDs, to enhance user security and access control. Secure your application while providing a convenient login experience for users.

https://bloggrammer.com/web-api-google-authentication-example

**Comment by dinhsacphd** [comment]
How to signin-google  with login action at this API Project at MVC .net core?

**Reply by bloggrammer** [reply]
It should work the same for MVC project. 

The `signin-google` is a callback function that gets called by Google

**Reply by dinhsacphd** [reply]
I create new MVC NET CORE project to Login API Project, But don't know how to get the response from API Project?
You have a solution, please support me.
Thanks a lot.

**Reply by bloggrammer** [reply]
Call the `Login` endpoint from your frontend. Look at step 6

**Reply by dinhsacphd** [reply]
Thanks. But After login, have a errror:
"InvalidOperationException: No authenticationScheme was specified, and there was no DefaultSignInScheme found. The default schemes can be set using either AddAuthentication(string defaultScheme) or AddAuthentication(Action<AuthenticationOptions> configureOptions)."

**Reply by dinhsacphd** [reply]
Edit the same your code: have a error
 Exception: An error was encountered while handling the remote login.
Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler<TOptions>.HandleRequestAsync()

**Reply by bloggrammer** [reply]
Hi, just reading your reply. 

Let me know if you still encounter this error.

Cheers!

**Comment by nithrous** [comment]
In google cloud console you've entered:
```https://yourdomain.com/signin-google```

But your actual route is:
```https://yourdomain.com/account/signin-google```

Fix please

**Comment by BongJaeJeong** [comment]
Hi, I tried my own asp.net web api project with google oauth with the information you provided.

1. I added this in program.cs
builder.Services.AddAuthentication(options =>
{
    options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
    options.DefaultChallengeScheme = GoogleDefaults.AuthenticationScheme;
})
.AddCookie()
.AddGoogle(options =>
{
    options.ClientId = appSettings.GoogleOAuth.ClientId;
    options.ClientSecret = appSettings.GoogleOAuth.ClientSecret;
    options.CallbackPath = "/api/auth/signin-google";
});

2. added redirecturi in google console --> http://localhost:44350/api/auth/signin-google

3. made the apicontroller Login + GoogleLogin exactly the same.

But when i call localhost:44350/api/auth/google-login, i keep getting error saying
: The oauth state was missing or invalid.

do you know why?
